module/zfs/dbuf.c

```diff
@@ -5005,6 +5040,8 @@ dbuf_write_children_ready(zio_t *zio, arc_buf_t *buf, void *vdb)
 	DB_DNODE_EXIT(db);
 	ASSERT3U(epbs, <, 31);
 
+	mutex_enter(&db->db_mtx);
+	rw_enter(&db->db_rwlock, RW_READER);
```

Comment:
I don't think we need `db_mtx` lock here, while `db_rwlock` not sure, since at this point all modifications must be completed, since the block is going to be written.  I think previous `db_rwlock` below should be sufficient.